### PR TITLE
upgrades,systemschema: add covering index to system.privileges on `path`

### LINKED ISSUE #93525: systemschema: add index on system.privileges(path)
**Is your feature request related to a problem? Please describe.**

This system table is basically always queried by filtering on the `path` column. There's no index on it, so all the queries do full scans. Luckily the table is probably going to be small (under a thousand rows), so this probably isn't _too_ bad.

**Describe the solution you'd like**

Add an index on `path`. (update the hardcoded schema and add an upgrade migration)

**Describe alternatives you've considered**

We should have made the primary key of the table be `(path, user)`, but right now it's `(user, path)`. We could ALTER the primary key instead of adding a different index.

**Additional context**

Realized during review of https://github.com/cockroachdb/cockroach/pull/93444

Jira issue: CRDB-22380

### PR DESCRIPTION
**Pull request by e-mbrown**
Resolves #93525

Queries to the `system.privileges` table are usually filtered by path. With a primary key using `(user,path)` all those queries were full scans.  This commit changes the primary key to be `(path,user)`.

Release note: None

### CODE REVIEW

File: pkg/sql/catalog/systemschema/system.go [hunk 1]
```diff
@@ -2921,9 +2921,9 @@ var (
 				Name:                "primary",
 				ID:                  1,
 				Unique:              true,
-				KeyColumnNames:      []string{"username", "path"},
```
Comment (andyyang890) [inline pkg/sql/catalog/systemschema/system.go:2924]: I think you'll probably need to write a migration that alters the schema for existing (22.2) clusters. The schema specified in this file is only used when bootstrapping a cluster for the first time.

**Review by e-mbrown — COMMENTED** [review]
Hmm, okie dokie. Ill ask Rafi if this is the direction we want to take, and start working on that migration.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/96460#-:-NNIx01tD968uewC1c3X:bqwjs70)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rafiss)

**Review by andyyang890 — CHANGES_REQUESTED** [review]
I think you'll need to rebase since I recently merged another PR that runs a migration on this table. Also I think you need to run `./dev gen bazel` to make sure the new files you're adding are being included in the `BUILD.bazel` file.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/96460#-:-NOBq0RU6N1TU7oWtf5Y:b-bwy20o)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @e-mbrown and @rafiss)

___
*[`-- commits` line 2 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/96460#-NOBpk26-ZY2z_xAVe3-):*
nit: "covering" is misspelled
___
*[`pkg/clusterversion/cockroach_versions.go` line 411 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/96460#-NOBs7tI6h-33FXCJDRK:-NOBs7tI6h-33FXCJDRL:b-82bbp3) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a8b5d64eef120f6767f13832cab0980ccd39609/pkg/clusterversion/cockroach_versions.go#L411)):*
> ```Go
> 	V23_1_KVDirectColumnarScans
> 
> 	// V23_1_AlterSystemPrivilegesAddIndex adds a covering secondary index to
> ```

nit: Can you update the comment to mention the columns the index is indexed on?
___
*[`pkg/clusterversion/cockroach_versions.go` line 413 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/96460#-NOBrnUC6I-uDWQntEug:-NOBrnUC6I-uDWQntEuh:b8oab9g) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a8b5d64eef120f6767f13832cab0980ccd39609/pkg/clusterversion/cockroach_versions.go#L413)):*
> ```Go
> 	// V23_1_AlterSystemPrivilegesAddIndex adds a covering secondary index to
> 	// system.privileges
> 	V23_1AlterSystemPrivilegesAddIndex
> ```

nit: Could you rename this to something more specific? Maybe `...AddIndexOnPathAndUsername`?
___
*[`pkg/sql/catalog/systemschema/system.go` line 737 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/96460#-NOBrcuoBHUgOq0q7HW-:-NOBrcuoBHUgOq0q7HW0:b-6ldqxr) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a8b5d64eef120f6767f13832cab0980ccd39609/pkg/sql/catalog/systemschema/system.go#L737)):*
> ```Go
> );`
> 
> 	SystemPrivilegeTableSchema = `
> ```

You need to add the new index here too
___
*[`pkg/sql/catalog/systemschema/system.go` line 2929 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/96460#-NOBsZeQCJ5NmRrjRCYC:-NOBsZeQCJ5NmRrjRCYD:b-2u4de1) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a8b5d64eef120f6767f13832cab0980ccd39609/pkg/sql/catalog/systemschema/system.go#L2929)):*
> ```Go
> 			},
> 			descpb.IndexDescriptor{
> 				Name:                "path_idx",
> ```

Once you update the schema like I mentioned above, this will probably become a more specific index name (likely `privileges_path_username_idx_key`)
___
*[`pkg/upgrade/upgrades/system_privileges_index_migration.go` line 41 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/96460#-NOBqwMp-38Fk0S_UHSd:-NOBqwMp-38Fk0S_UHSe:bah0m7w) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a8b5d64eef120f6767f13832cab0980ccd39609/pkg/upgrade/upgrades/system_privileges_index_migration.go#L41)):*
> ```Go
> 		name:           "add-secondary-index-system-priv",
> 		schemaList:     []string{"path_idx"},
> 		query:          `CREATE INDEX IF NOT EXISTS path_idx ON system.privileges (path,username) STORING (privileges, grant_options)`,
> ```

This should probably be a unique index.
___
*[`pkg/upgrade/upgrades/upgrades.go` line 262 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/96460#-NOASY1O-BLSjOcnQs0z:-NOASY1O-BLSjOcnQs1-:byf2qg) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a8b5d64eef120f6767f13832cab0980ccd39609/pkg/upgrade/upgrades/upgrades.go#L262)):*
> ```Go
> 	),
> 	upgrade.NewTenantUpgrade(
> 		"add secondary index to table system.privileges",
> ```

nit: can you change this to be slightly more specific? maybe something like "add secondary index on (path, username) to system.privileges table"
___
*[`pkg/sql/opt/exec/execbuilder/testdata/explain_analyze` line 136 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/96460#-NOBr89RExwKxhxVuyS2:-NOBr89RExwKxhxVuyS3:bfl6u0l) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a8b5d64eef120f6767f13832cab0980ccd39609/pkg/sql/opt/exec/execbuilder/testdata/explain_analyze#L136)):*
> ```
>   columns: (username, path, privileges, grant_options)
>   estimated row count: 10 (missing stats)
>   table: privileges@primary
> ```

Do you know why the new index isn't being used here?
___
*[`pkg/upgrade/upgrades/system_privileges_index_migration_test.go` line 17 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/96460#-NOBrMt49OJk3RIAHr1g:-NOBrMt49OJk3RIAHr1h:b4fwno0) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a8b5d64eef120f6767f13832cab0980ccd39609/pkg/upgrade/upgrades/system_privileges_index_migration_test.go#L17)):*
> ```Go
> )
> 
> func TestSystemIndexMigration(t *testing.T) {
> ```

nit: Can you name this something more specific? Maybe at least include the table name 
___
*[`pkg/upgrade/upgrades/system_privileges_index_migration_test.go` line 33 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/96460#-NOBtVCF-3ferTFw3TxX:-NOBtVCGD9b2a7pVSRjp:be16pld) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a8b5d64eef120f6767f13832cab0980ccd39609/pkg/upgrade/upgrades/system_privileges_index_migration_test.go#L33)):*
> ```Go
> 				Server: &server.TestingKnobs{
> 					DisableAutomaticVersionUpgrade: make(chan struct{}),
> 					BinaryVersionOverride:          clusterversion.ByKey(clusterversion.V23_1AlterSystemPrivilegesAddIndex - 1),
> ```

There doesn't seem to be any code in this test that actually runs the migration. You'll probably want to hardcode the old descriptor and use `upgrades.InjectLegacyTable` before running the migration

**Review by andyyang890 — CHANGES_REQUESTED** [review]
Thanks for making the changes! Left a few more small comments. Also, it looks like you'll need to update (manually copy and replace) the hash values in `pkg/sql/catalog/bootstrap/testdata/testdata` to make CI pass.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/96460#-:-NOGV1yc5YZ8Xyx0EVWz:b54ed68)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @e-mbrown, @msirek, and @rafiss)

___
*[`-- commits` line 2 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/96460#-NOBpk26-ZY2z_xAVe3-):*
<details><summary><i>Previously, andyyang890 (Andy Yang) wrote…</i></summary><blockquote>

nit: "covering" is misspelled
</blockquote></details>

nit: generally, the word following the package name should not be capitalized. can you also update the PR title to match your new commit message?
___
*[`pkg/upgrade/upgrades/system_privileges_index_migration.go` line 41 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/96460#-NOBqwMp-38Fk0S_UHSd:-NOGTWPdEHlsu5P9DCEp:b-w1u4s5) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a8b5d64eef120f6767f13832cab0980ccd39609/pkg/upgrade/upgrades/system_privileges_index_migration.go#L41)):*
<details><summary><i>Previously, andyyang890 (Andy Yang) wrote…</i></summary><blockquote>

This should probably be a unique index.
</blockquote></details>

`s/path_idx/privileges_path_username_idx_key/` on this line and the line above
___
*[`pkg/upgrade/upgrades/system_privileges_index_migration_test.go` line 98 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/96460#-NOGUChgCcXjv5zp7Sox:-NOGUChh9fVvNPlne6OO:bvtsx7m) ([raw file](https://github.com/cockroachdb/cockroach/blob/59c420bdbbe82965dce884539e1d749d07ce6dba/pkg/upgrade/upgrades/system_privileges_index_migration_test.go#L98)):*
> ```Go
> 
> func getTableDescForSystemPrivilegesTableBeforeIndexonPathUsername() *descpb.TableDescriptor {
> 	return &descpb.TableDescriptor{
> ```

You should use the latest descriptor before your changes (i.e. the one that has the `user_id` column already) since that is the one that most closely matches the state of the table in actual databases before your upgrade runs.

**Review by msirek — COMMENTED** [review]
The changes look good from a SQL Queries perspective. 

> Reviewed 11 of 12 files at r3, 6 of 12 files at r5, 1 of 2 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/96460#-:-NORBvr74-xWYVG1GJWU:b-5doq00)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andyyang890, @e-mbrown, and @rafiss)

**Review by andyyang890 — CHANGES_REQUESTED** [review]
Looks good other than one comment about the old descriptor in the unit test and needing a test rewrite for `TestValidateSystemSchemaAfterBootStrap`

> Reviewed 1 of 12 files at r5, 1 of 1 files at r7, 8 of 8 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/96460#-:-NOVuPbJ7BQH-v6ZpJQL:b5bykvp)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @e-mbrown and @rafiss)

___
*[`pkg/upgrade/upgrades/system_privileges_index_migration_test.go` line 98 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/96460#-NOGUChgCcXjv5zp7Sox:-NOVu7MR8VUWc4f0ori8:bju7j5f) ([raw file](https://github.com/cockroachdb/cockroach/blob/59c420bdbbe82965dce884539e1d749d07ce6dba/pkg/upgrade/upgrades/system_privileges_index_migration_test.go#L98)):*
<details><summary><i>Previously, andyyang890 (Andy Yang) wrote…</i></summary><blockquote>

You should use the latest descriptor before your changes (i.e. the one that has the `user_id` column already) since that is the one that most closely matches the state of the table in actual databases before your upgrade runs.
</blockquote></details>

Can you add the `privileges_path_user_id_key` in too? You'll probably need to update `expectedSchema` above after too.

**Review by e-mbrown — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/96460)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andyyang890 and @rafiss)

___
*[`pkg/sql/catalog/systemschema/system.go` line 737 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/96460#-NOBrcuoBHUgOq0q7HW-:-NOFHm_u2qkp3i9rhLvM:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a8b5d64eef120f6767f13832cab0980ccd39609/pkg/sql/catalog/systemschema/system.go#L737)):*
<details><summary><i>Previously, andyyang890 (Andy Yang) wrote…</i></summary><blockquote>

You need to add the new index here too
</blockquote></details>

Done.
___
*[`pkg/sql/catalog/systemschema/system.go` line 2929 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/96460#-NOBsZeQCJ5NmRrjRCYC:-NOFHw8ZC-M2FR3tnqrY:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a8b5d64eef120f6767f13832cab0980ccd39609/pkg/sql/catalog/systemschema/system.go#L2929)):*
<details><summary><i>Previously, andyyang890 (Andy Yang) wrote…</i></summary><blockquote>

Once you update the schema like I mentioned above, this will probably become a more specific index name (likely `privileges_path_username_idx_key`)
</blockquote></details>

Done.
___
*[`pkg/upgrade/upgrades/system_privileges_index_migration.go` line 41 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/96460#-NOBqwMp-38Fk0S_UHSd:-NOFIFGq3RclW3eCj8kc:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a8b5d64eef120f6767f13832cab0980ccd39609/pkg/upgrade/upgrades/system_privileges_index_migration.go#L41)):*
<details><summary><i>Previously, andyyang890 (Andy Yang) wrote…</i></summary><blockquote>

This should probably be a unique index.
</blockquote></details>

Done.
___
*[`pkg/sql/opt/exec/execbuilder/testdata/explain_analyze` line 136 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/96460#-NOBr89RExwKxhxVuyS2:-NOFIRr02qH9sA-ibLpA:b99qk5b) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a8b5d64eef120f6767f13832cab0980ccd39609/pkg/sql/opt/exec/execbuilder/testdata/explain_analyze#L136)):*
<details><summary><i>Previously, andyyang890 (Andy Yang) wrote…</i></summary><blockquote>

Do you know why the new index isn't being used here?
</blockquote></details>

I didnt rewrite the test lol. Good catch
___
*[`pkg/upgrade/upgrades/system_privileges_index_migration_test.go` line 98 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/96460#-NOGUChgCcXjv5zp7Sox:-NOLSjh44qJD8Ictoygx:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/59c420bdbbe82965dce884539e1d749d07ce6dba/pkg/upgrade/upgrades/system_privileges_index_migration_test.go#L98)):*
<details><summary><i>Previously, andyyang890 (Andy Yang) wrote…</i></summary><blockquote>

You should use the latest descriptor before your changes (i.e. the one that has the `user_id` column already) since that is the one that most closely matches the state of the table in actual databases before your upgrade runs.
</blockquote></details>

Done.

**Review by andyyang890 — APPROVED** [review]
:lgtm:

needs another crlfmt to make the linter happy but otherwise looks good!

> Reviewed 1 of 2 files at r10, 1 of 1 files at r11, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/96460#-:-NOWAgvlAV75wK8xAJQU:b-r5en9k)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @rafiss)

**Comment by e-mbrown** [comment]
Thanks for the help and reviews!

bors r=andyyang890

**Pull request merged**